src/string-builder.js

```diff
@@ -18,11 +39,28 @@ StringBuilder.prototype.cat = function(){
                 concat(...element);
             } else if (typeof element === 'function') {
                 concat(element());
+            } else if (decorators && decorators.length > 0) {
+                concatWithDecorators(decorators, element);
             } else {
                 buffer.push(element);
             }
         }
     }
+
+    function concatWithDecorators(decorators, element) {
+        var prefixes = [];
+        var suffixes = [];
+        
+        for (let i = decorators.length - 1; i >= 0; i--) {
+            let decorator = decorators[i];
+            decorator.execute(prefixes, suffixes);
+        }
+
+        cat.call({ buffer : buffer}, prefixes.reverse());
```

Comment:
Take a look of the `Array.unshift` to avoid calling `.reverse()` each time we need to push the prefixes.